### PR TITLE
Merge small metadata files of DMFile

### LINKED ISSUE #6825: Merge small metadata files of DMFile into one.
## Enhancement

In order to reduce the number of I/O, the four metadata files -- config, meta.txt, property, pack under the DMFile directory are merged into one -- meta.

### PR DESCRIPTION
**Pull request by JinheLin**
### What problem does this PR solve?

Issue Number: close #6825 

Problem Summary:
In order to reduce the number of I/O, the four metadata files -- config, meta.txt, property, pack under the DMFile directory are merged into one -- meta.

### What is changed and how it works?
New metadata file format:
```
 |Pack Stats|Pack Properties|Column Stats|Pack Stats Handle|Pack Properties Handle|Column Stats Handle|Meta Block Handle Count|DMFile Version|Checksum|MetaFooter|
 |----------------------------------------Checksum include-----------------------------------------------------------------------------------|
```
-  `MetaFooter` is saved at the end of the file, with fixed length, it contains checksum algorithm and checksum frame length.
- First, read `MetaFooter` and `Checksum`, and check data integrity.
- Second, parse handle and parse corresponding data.
- `PackStatsHandle`, `PackPropertiesHandle` and `ColumnStatsHandle` are offset and size of `PackStats`, `PackProperties` and `ColumnStats`.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
  - Enable metav2 is dtworkload and run `./tiflash dtworkload --testing_type=daily_perf`. `dtworkload` will read, write and restore   the storage engine continuously. 
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Storages/DeltaMerge/File/DMFile.cpp [hunk 1]
```diff
@@ -701,5 +757,165 @@ void DMFile::initializeIndices()
...
+{
+    auto rbuf = createReadBufferFromFileBaseByFileProvider(
+        file_provider,
+        metav2Path(),
+        encryptionMetav2Path(),
+        meta_checksum_frame_length,
+        /*read_limiter*/ nullptr,
+        meta_checksum_algorithm,
+        meta_checksum_frame_length);
+    std::vector<char> buf(meta_checksum_frame_length);
```
[the remark this answers is not in the document]
Comment (lidezhu) [reply dbms/src/Storages/DeltaMerge/File/DMFile.cpp:825]: This is ok for me.

File: dbms/src/Storages/DeltaMerge/File/DMFile.h [hunk 2]
```diff
@@ -126,21 +126,64 @@ class DMFile : private boost::noncopyable
...
-        {}
         UInt64 offset;
         UInt64 size;
     };
-    using SubFileStats = std::unordered_map<String, SubFileStat>;
+    static_assert(std::is_standard_layout_v<MetaBlockHandle> && sizeof(MetaBlockHandle) == sizeof(UInt64) * 2);
+
+    struct MetaFooter
+    {
+        UInt64 meta_fromat_version;
```
[the remark this answers is not in the document]
Comment (JaySon-Huang) [reply dbms/src/Storages/DeltaMerge/File/DMFile.h:181]: +1

File: dbms/src/Storages/DeltaMerge/File/DMFileWriter.h [hunk 3]
```diff
@@ -159,18 +159,26 @@ class DMFileWriter
...
     ColumnStreams column_streams;
 
-    WriteBufferFromFileBasePtr pack_stat_file;
-
     FileProviderPtr file_provider;
     WriteLimiterPtr write_limiter;
 
+    // If dmfile->useMetaV2() is true, `file` is for metav2,
+    // else `file` is for pack stats.
+    WriteBufferFromFileBasePtr file;
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/File/DMFileWriter.h:180]: meta_file maybe a better name

File: dbms/src/Storages/DeltaMerge/File/DMFileWriter.h [hunk 4]
```diff
@@ -159,18 +159,26 @@ class DMFileWriter
     /// FileNameBase -> Stream.
     void addStreams(ColId col_id, DataTypePtr type, bool do_index);
 
+    WriteBufferFromFileBasePtr createFile();
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/File/DMFileWriter.h:162]: createMetaFile()

File: dbms/src/Storages/DeltaMerge/File/DMFileWriter.cpp [hunk 5]
```diff
@@ -125,16 +151,40 @@ void DMFileWriter::write(const Block & block, const BlockProperty & block_proper
...
+    {
+        finalizeMetaV2();
+    }
+    else
+    {
+        finalizeMeta();
+    }
+}
 
+void DMFileWriter::finalizeMeta()
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/File/DMFileWriter.cpp:168]:
```suggestion
void DMFileWriter::finalizeMetaV1()
```

File: dbms/src/Storages/DeltaMerge/File/DMFileWriter.cpp [hunk 6]
```diff
@@ -125,16 +151,40 @@ void DMFileWriter::write(const Block & block, const BlockProperty & block_proper
...
 {
-    pack_stat_file->sync();
-
     for (auto & cd : write_columns)
     {
         finalizeColumn(cd.id, cd.type);
     }
+    if (dmfile->useMetaV2())
+    {
+        finalizeMetaV2();
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/File/DMFileWriter.cpp:160]:
```suggestion
        // Some fields of ColumnStat is set in `finalizeColumn `, must call finalizeMetaV2 after all column finalized
        finalizeMetaV2();
```

**Review by lidezhu — APPROVED** [review]
LGTM

File: dbms/src/Storages/FormatVersion.h [hunk 7]
```diff
@@ -39,6 +39,7 @@ using Version = UInt32;
 inline static constexpr Version V0 = 0;
 inline static constexpr Version V1 = 1; // Add column stats
 inline static constexpr Version V2 = 2; // Add checksum and configuration
+inline static constexpr Version V3 = 3; // Use Meta V2
```
Comment (JaySon-Huang) [inline dbms/src/Storages/FormatVersion.h:42]: Need to set STORAGE_FORMAT_V5.dm_file = DMFileFormat::V3;
 
https://github.com/pingcap/tiflash/blob/f327b7fd350def2719fb0c3b6ed30af290b1e9d4/dbms/src/Storages/FormatVersion.h#L124-L131

File: dbms/src/Storages/DeltaMerge/File/DMFile.cpp [hunk 8]
```diff
@@ -101,14 +102,15 @@ String DMFile::ngcPath() const
     return getNGCPath(parent_path, file_id, status);
 }
 
-DMFilePtr DMFile::create(UInt64 file_id, const String & parent_path, DMConfigurationOpt configuration)
+DMFilePtr DMFile::create(UInt64 file_id, const String & parent_path, DMConfigurationOpt configuration, bool use_meta_v2_)
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/File/DMFile.cpp:105]:
```suggestion
DMFilePtr DMFile::create(UInt64 file_id, const String & parent_path, DMConfigurationOpt configuration, DMFileFormat::Version version = STORAGE_FORMAT_CURRENT.dm_file)
```

File: dbms/src/Storages/DeltaMerge/File/DMFile.h [hunk 9]
```diff
@@ -282,18 +325,24 @@ class DMFile : private boost::noncopyable
...
+#else
+public:
+#endif
     DMFile(UInt64 file_id_,
            UInt64 page_id_,
            String parent_path_,
            Status status_,
-           DMConfigurationOpt configuration_ = std::nullopt)
+           DMConfigurationOpt configuration_ = std::nullopt,
+           bool use_meta_v2_ = false)
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/File/DMFile.h:338]:
```
    DMFile(UInt64 file_id_,
           UInt64 page_id_,
           String parent_path_,
           Status status_,
           DMConfigurationOpt configuration_ = std::nullopt)
           DMConfigurationOpt configuration_ = std::nullopt,
           DMFileFormat::Version version = STORAGE_FORMAT_CURRENT.dm_file)
```

File: dbms/src/Storages/DeltaMerge/File/DMFile.h [hunk 10]
```diff
@@ -282,18 +325,24 @@ class DMFile : private boost::noncopyable
...
         : file_id(file_id_)
         , page_id(page_id_)
         , parent_path(std::move(parent_path_))
         , status(status_)
         , configuration(std::move(configuration_))
         , log(Logger::get())
+        , use_meta_v2(use_meta_v2_)
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/File/DMFile.h:345]:
```suggestion
        , use_meta_v2(version == DMFileFormat::V3)
```

File: dbms/src/Storages/DeltaMerge/tests/gtest_dm_file.cpp [hunk 11]
```diff
@@ -222,6 +221,112 @@ try
...
+    std::vector<DMFileBlockOutputStream::BlockProperty> block_propertys;
+    block_propertys.push_back(block_property1);
+    block_propertys.push_back(block_property2);
+    DMFilePtr dmfile1, dmfile2;
+    {
+        Block block1 = DMTestEnv::prepareSimpleWriteBlock(0, num_rows_write / 2, false);
+        Block block2 = DMTestEnv::prepareSimpleWriteBlock(num_rows_write / 2, num_rows_write, false);
+        auto mode = DMFileMode::DirectoryChecksum;
+        auto configuration = createConfiguration(mode);
+        dmfile1 = DMFile::create(1, parent_path, std::move(configuration), /*use_meta_v2*/ false);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/tests/gtest_dm_file.cpp:301]:
```suggestion
        dmfile1 = DMFile::create(1, parent_path, std::move(configuration), /*version*/ DMFileFormat::V2);
```

File: dbms/src/Storages/DeltaMerge/tests/gtest_dm_file.cpp [hunk 12]
```diff
@@ -222,6 +221,112 @@ try
...
+        stream->write(block1, block_property1);
+        stream->write(block2, block_property2);
+        stream->writeSuffix();
+    }
+    {
+        Block block1 = DMTestEnv::prepareSimpleWriteBlock(0, num_rows_write / 2, false);
+        Block block2 = DMTestEnv::prepareSimpleWriteBlock(num_rows_write / 2, num_rows_write, false);
+        auto mode = DMFileMode::DirectoryMetaV2;
+        auto configuration = createConfiguration(mode);
+        dmfile2 = DMFile::create(1, parent_path, std::move(configuration), /*use_meta_v2*/ true);
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/tests/gtest_dm_file.cpp:313]:
```suggestion
        dmfile2 = DMFile::create(1, parent_path, std::move(configuration), /*version*/ DMFileFormat::V3);
```

File: dbms/src/Storages/DeltaMerge/File/DMFile.cpp [hunk 13]
```diff
@@ -701,5 +760,205 @@ void DMFile::initializeIndices()
...
+    {
+        ptr = ptr - sizeof(MetaBlockHandle);
+        const auto * handle = reinterpret_cast<const MetaBlockHandle *>(ptr);
+        parsePackProperty(buffer.substr(handle->offset, handle->size));
+    }
+    {
+        ptr = ptr - sizeof(MetaBlockHandle);
+        const auto * handle = reinterpret_cast<const MetaBlockHandle *>(ptr);
+        parsePackStat(buffer.substr(handle->offset, handle->size));
+    }
```
Comment (JaySon-Huang) [inline dbms/src/Storages/DeltaMerge/File/DMFile.cpp:909]: Maybe add a `UInt64 block_type` before each MetaBlockHandle to indicate which type this block is?
Adding this we can add more type of optional "Block" in the future without changing the file format.

**Review by JaySon-Huang — APPROVED** [review]
LGTM

**Pull request merged**